hibernate-core/src/main/java/org/hibernate/query/hql/internal/SemanticQueryBuilder.java

```diff
@@ -3653,14 +3657,31 @@ public Object visitStrFunction(HqlParser.StrFunctionContext ctx) {
 				? new SqmDistinct<>( arg, getCreationContext().getNodeBuilder() )
 				: arg;
 
-		return getFunctionDescriptor("count").generateSqmExpression(
-				argument,
+		return getFunctionDescriptor("count").generateAggregateSqmExpression(
+				singletonList( argument ),
+				getFilterExpression( ctx.filterClause() ),
 				resolveExpressableTypeBasic( Long.class ),
 				creationContext.getQueryEngine(),
 				creationContext.getJpaMetamodel().getTypeConfiguration()
 		);
 	}
 
+	private SqmPredicate getFilterExpression(HqlParser.FilterClauseContext filterClauseCtx) {
+		if (filterClauseCtx == null) {
+			return null;
+		}
+		return (SqmPredicate) filterClauseCtx.whereClause().predicate().accept( this );
+	}
+
+	@Override
+	public SqmExpression<?> visitFilterClause(HqlParser.FilterClauseContext ctx) {
```

Comment:
You can remove this as I think you handled all uses above.

Reply:
Isn't this needed in case 'someone' were to want to walk the tree and drill down into the filter clause?

Reply:
Every use of the rule is handled explicitly. This rule is only invoked if the default handler for the AST node is visited, which is never the case.